src/main/java/ch/njol/skript/expressions/ExprInput.java

```diff
@@ -0,0 +1,163 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.expressions;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.classes.ClassInfo;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.ExpressionType;
+import ch.njol.skript.lang.InputSource;
+import ch.njol.skript.lang.InputSource.InputData;
+import ch.njol.skript.lang.Literal;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.skript.registrations.Classes;
+import ch.njol.skript.util.Utils;
+import ch.njol.util.Kleenean;
+import org.bukkit.event.Event;
+import org.eclipse.jdt.annotation.Nullable;
+import org.skriptlang.skript.lang.converter.Converters;
+
+import java.lang.reflect.Array;
+import java.util.Set;
+
+@Name("Input")
+@Description({
+	"Represents the input in a filter expression or sort effect.",
+	"For example, if you ran 'broadcast \"something\" and \"something else\" where [input is \"something\"]",
+	"the condition would be checked twice, using \"something\" and \"something else\" as the inputs.",
+	"The 'input index' pattern can be used when acting on a variable to access the index of the input."
+})
+@Examples({
+	"send \"congrats on being staff!\" to all players where [input has permission \"staff\"]",
+	"sort {_list::*} based on length of input index"
+})
+@Since("2.2-dev36, INSERT_VERSION (input index)")
+public class ExprInput<T> extends SimpleExpression<T> {
+
+	static {
+		Skript.registerExpression(ExprInput.class, Object.class, ExpressionType.COMBINED,
+			"input",
+			"%*classinfo% input",
```

Comment:
could you please add an error if a plural classinfo is used? should be easy with ClassInfoReference

Reply:
pikachu are you allergic to making a single review

Reply:
![68747470733a2f2f6d65646961312e74656e6f722e636f6d2f6d2f34685f516b4f50574c395941414141642f6361742d736e65657a652e676966](https://github.com/SkriptLang/Skript/assets/28607612/7ee444fb-82b7-4461-95d8-933863a02170)